### PR TITLE
RESTWS-880: Add support to only return published forms from the backend

### PR DESCRIPTION
**Pull request by abertnamanya**
## Description of what I changed

- Created  FormSearchHandler class
- Added formTestDataSet.xml file in the test resource
- Created FormSearchHandlerTest class and carried out units on the search parameters 

## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-880

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.openmrs.module.webservices.rest.web.resource.impl.EmptySearchResult;
+import org.openmrs.module.webservices.rest.web.resource.impl.NeedsPaging;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FormSearchHandler1_8 implements SearchHandler {
+
+	private final SearchConfig searchConfig = new SearchConfig("default", RestConstants.VERSION_1 + "/form", Arrays.asList(
+			"1.8.*", "1.9.*", "1.10.*", "1.11.*", "1.12.*", "2.0.*", "2.1.*", "2.2.*", "2.3.*", "2.4.*", "2.5.*"),
```
Comment (corneliouzbett) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:32]: Use range instead; `Collections.singletonList("1.8 - 2.*")`

Comment (abertnamanya) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java]: Thank you @corneliouzbett, Please check and review Just made the changes

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FormSearchHandler1_8 implements SearchHandler {
+
+	private final SearchConfig searchConfig = new SearchConfig("default", RestConstants.VERSION_1 + "/form", Collections.singletonList("1.8 - 2.*"),
+
+			new SearchQuery.Builder(
+					"Allows you to find form by published and unpublished status")
```
Comment (dkayiwa) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:35]: The above sounds as if two parameters are used. One for published and the other for unpublished.

File: omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		return formService.getAllForms(false).size();
+	}
+
+	/**
+	 * @verifies return location by tag uuid
+	 * @see FormSearchHandler1_8#getSearchConfig()
+	 */
+	@Test
+	public void getSearchConfig_shouldReturnFormByPublishedStatus() throws Exception {
+		MockHttpServletRequest req = request(RequestMethod.GET, getURI());
```
Comment (dkayiwa) [inline omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandlerTest.java:58]: Do you wanna add a test where `published` is `false`?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+					.withRequiredParameters("published").build());
+
+	@Override
+	public SearchConfig getSearchConfig() {
+		return searchConfig;
+	}
+
+	@Override
+	public PageableResult search(RequestContext context) throws ResponseException {
+		boolean formStatus = Boolean.parseBoolean(context.getParameter("published"));
```
Comment (dkayiwa) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:45]: Did you intentionally leave out some sort of name parameter?

Comment (dkayiwa) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:45]: Is there a possibility of any of the search parameters not being passed and hence null?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java (hunk 2 above)
Comment (abertnamanya) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:35]: @dkayiwa  Basically one passes one parameter which is published with a value either true or false, But I could only limit it to published

Comment (corneliouzbett) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:35]: @abertnamanya Daniel is referring to the message, it's unclear(suggests two parameters published and unpublished), consider changing to a more clear message, something like `Allows you to filter forms by published status`

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java (hunk 4 above)
Comment (corneliouzbett) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:45]:
```java
String publishedStatus = context.getParameter("published");
if (publishedStatus != null && !publishedStatus.isEmpty()) {
  boolean formStatus = Boolean.parseBoolean(publishedStatus);
  //continue the logic
}
//Handle when the published status isn't set then probably return an Empty search results
```

Comment (abertnamanya) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandler1_8.java:45]: Thank you @corneliouzbett

File: omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandlerTest.java (hunk 3 above)
Comment (abertnamanya) [reply omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandlerTest.java:58]: Yeah I have added the new test method

**Review by corneliouzbett — APPROVED** [review]
Thanks, @abertnamanya for working on this! Minor change otherwise looks good

File: omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	@Test
+	public void getSearchConfig_shouldReturnPublishedForms() throws Exception {
+		MockHttpServletRequest req = request(RequestMethod.GET, getURI());
+		req.addParameter("published", "true");
+		SimpleObject result = deserialize(handle(req));
+		List<Object> hits = result.get("results");
+		Assert.assertEquals(3, hits.size());
+	}
+	@Test
+	public void GetSearchConfig_ShouldReturnUnpublishedForms() throws Exception{
```
Comment (corneliouzbett) [inline omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/FormSearchHandlerTest.java:65]: Use camelCase for method naming
```suggestion
	public void getSearchConfig_shouldReturnUnpublishedForms() throws Exception{
```

**Pull request merged**